File: packages/crawler-go/internal/extractors/html_markup.go
package extractors

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/shroodler/crawler-go/internal/models"
	"golang.org/x/net/html"
)

func LoadCommentKeywords() []string {
	p := filepath.Join(RepoRoot(), "packages", "secret-patterns", "keywords", "html-comments.txt")
	b, err := os.ReadFile(p)
	if err != nil {
		return nil
	}
	var out []string
	for _, line := range strings.Split(string(b), "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		out = append(out, line)
	}
	return out
}

func commentHits(text string, keywords []string) bool {
	low := strings.ToLower(text)
	for _, key := range keywords {
		if key != "" && strings.Contains(low, strings.ToLower(key)) {
			return true
		}
	}
	return false
}

func snippet(text string, limit int) string {
	compact := strings.Join(strings.Fields(text), " ")
	if len(compact) > limit {
		return compact[:limit]
	}
	return compact
}

func ExtractHTMLComments(body, pageURL string, rules []Rule) []models.Finding {
	if body == "" {
		return nil
	}
	keywords := LoadCommentKeywords()
	if len(keywords) == 0 {
		return nil
	}
	doc, err := html.Parse(strings.NewReader(body))
	if err != nil {
		return nil
	}
	var findings []models.Finding
	seen := map[string]bool{}
	var walk func(*html.Node)
	walk = func(n *html.Node) {
		if n.Type == html.CommentNode {
			text := strings.TrimSpace(n.Data)
			if text != "" && commentHits(text, keywords) {
				key := strings.ToLower(text)
				if !seen[key] {
					seen[key] = true
					cat := "verbose-error"
					if len(ScanSecrets(text, pageURL, rules)) > 0 {
						cat = "secret"
					}
					ev := snippet(text, 80)
					findings = append(findings, models.Finding{
						ID: "html-comment", Severity: "info", Category: cat,
						URL: pageURL, Description: "HTML comment contains leftover TODO or credential-like text",
						Evidence: &ev,
					})
				}
			}
		}
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			walk(c)
		}
	}
	walk(doc)
	return findings
}

func ExtractMetaGenerator(body, pageURL string) []models.Finding {
	if body == "" {
		return nil
	}
	doc, err := html.Parse(strings.NewReader(body))
	if err != nil {
		return nil
	}
	var findings []models.Finding
	seen := map[string]bool{}
	var walk func(*html.Node)
	walk = func(n *html.Node) {
		if n.Type == html.ElementNode && n.Data == "meta" && strings.EqualFold(attr(n, "name"), "generator") {
			content := strings.TrimSpace(attr(n, "content"))
			if content != "" {
				key := strings.ToLower(content)
				if !seen[key] {
					seen[key] = true
					ev := snippet(content, 120)
					findings = append(findings, models.Finding{
						ID: "meta-generator", Severity: "info", Category: "header",
						URL: pageURL, Description: "meta generator tag discloses the application stack",
						Evidence: &ev,
					})
				}
			}
		}
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			walk(c)
		}
	}
	walk(doc)
	return findings
}

func ExtractHTMLMarkup(body, pageURL string, rules []Rule) []models.Finding {
	var out []models.Finding
	out = append(out, ExtractHTMLComments(body, pageURL, rules)...)
	out = append(out, ExtractMetaGenerator(body, pageURL)...)
	return out
}

File: packages/crawler-py/shroodler/extractors/html_markup.py
from __future__ import annotations

from functools import lru_cache
from pathlib import Path

from bs4 import BeautifulSoup, Comment

from shroodler.extractors.secrets import scan_text
from shroodler.models import Finding


def keywords_path() -> Path:
    here = Path(__file__).resolve()
    for parent in here.parents:
        candidate = parent / "packages" / "secret-patterns" / "keywords" / "html-comments.txt"
        if candidate.is_file():
            return candidate
        alt = parent / "secret-patterns" / "keywords" / "html-comments.txt"
        if alt.is_file():
            return alt
    raise FileNotFoundError("html-comments.txt not found")


def _parse_keywords(text: str) -> list[str]:
    out: list[str] = []
    for raw in text.splitlines():
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        out.append(line)
    return out


@lru_cache(maxsize=1)
def load_comment_keywords() -> list[str]:
    return _parse_keywords(keywords_path().read_text(encoding="utf-8"))


def _comment_hits(text: str, keywords: list[str]) -> bool:
    lowered = text.lower()
    return any(key.lower() in lowered for key in keywords)


def _snippet(text: str, limit: int = 80) -> str:
    compact = " ".join(text.split())
    if len(compact) > limit:
        return compact[:limit]
    return compact


def extract_html_comments(body: str, page_url: str) -> list[Finding]:
    if not body:
        return []
    keywords = load_comment_keywords()
    if not keywords:
        return []
    soup = BeautifulSoup(body, "lxml")
    findings: list[Finding] = []
    seen: set[str] = set()
    for node in soup.find_all(string=lambda t: isinstance(t, Comment)):
        text = str(node).strip()
        if not text or not _comment_hits(text, keywords):
            continue
        key = text.casefold()
        if key in seen:
            continue
        seen.add(key)
        secrets = scan_text(text, page_url)
        category = "secret" if secrets else "verbose-error"
        findings.append(
            Finding(
                id="html-comment",
                severity="info",
                category=category,
                url=page_url,
                description="HTML comment contains leftover TODO or credential-like text",
                evidence=_snippet(text),
            )
        )
    return findings


def extract_meta_generator(body: str, page_url: str) -> list[Finding]:
    if not body:
        return []
    soup = BeautifulSoup(body, "lxml")
    findings: list[Finding] = []
    seen: set[str] = set()
    for tag in soup.find_all("meta"):
        name = tag.get("name")
        if not name or str(name).lower() != "generator":
            continue
        content = (tag.get("content") or "").strip()
        if not content or content.casefold() in seen:
            continue
        seen.add(content.casefold())
        findings.append(
            Finding(
                id="meta-generator",
                severity="info",
                category="header",
                url=page_url,
                description="meta generator tag discloses the application stack",
                evidence=_snippet(content, 120),
            )
        )
    return findings


def extract_html_markup(body: str, page_url: str) -> list[Finding]:
    return extract_html_comments(body, page_url) + extract_meta_generator(body, page_url)
